File: backend/demo/src/test/java/ControllerTest/UploadControllerTest.java
package ControllerTest;

import com.prospektai.demo.PdfAiApplication;
import com.prospektai.demo.controller.UploadController;
import com.prospektai.demo.model.OfferData;
import com.prospektai.demo.repository.OfferDataRepository;
import com.prospektai.demo.service.PdfProcessingService;
import org.junit.jupiter.api.Test;
import org.mockito.Mock;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.autoconfigure.web.servlet.WebMvcTest;
import org.springframework.http.MediaType;
import org.springframework.mock.web.MockMultipartFile;
import org.springframework.test.context.ContextConfiguration;
import org.springframework.test.web.servlet.MockMvc;

import java.util.List;

import static org.mockito.Mockito.when;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.*;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.*;

@WebMvcTest(UploadController.class)
@ContextConfiguration(classes = PdfAiApplication.class)
class UploadControllerTest {
    @Autowired
    private MockMvc mockMvc;

    @Mock
    private PdfProcessingService pdfProcessingService;

    @Mock
    private OfferDataRepository offerDataRepository;

    @Test
    void uploadFile_shouldReturnSuccess() throws Exception {
        MockMultipartFile file = new MockMultipartFile(
                "file",
                "test.pdf",
                MediaType.APPLICATION_PDF_VALUE,
                "PDF content".getBytes()
        );

        mockMvc.perform(multipart("/api/upload")
                        .file(file)
                        .param("pagesPerChunk", "3"))
                .andExpect(status().isOk());
    }

    @Test
    void getAllOffers_shouldReturnOffers() throws Exception {
        OfferData offer = OfferData.builder()
                .id(1L)
                .productName("Test Produkt")
                .price("9.99")
                .build();

        when(offerDataRepository.findAll()).thenReturn(List.of(offer));

        mockMvc.perform(get("/api/offers"))
                .andExpect(status().isOk())
                .andExpect(jsonPath("$[0].productName").value("Test Produkt"));
    }
}

File: backend/demo/src/test/kotlin/com/prospektai/demo/controller/UploadControllerTest.kt
package com.prospektai.demo.controller

import com.prospektai.demo.model.OfferData
import com.prospektai.demo.repository.OfferDataRepository
import com.prospektai.demo.service.PdfProcessingService
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.mockito.Mock
import org.mockito.InjectMocks
import org.mockito.MockitoAnnotations
import org.mockito.Mockito.*
import org.springframework.http.MediaType
import org.springframework.mock.web.MockMultipartFile
import org.springframework.test.web.servlet.MockMvc
import org.springframework.test.web.servlet.setup.MockMvcBuilders
import org.springframework.test.web.servlet.request.MockMvcRequestBuilders.*
import org.springframework.test.web.servlet.result.MockMvcResultMatchers.*

class UploadControllerTest {

    private lateinit var mockMvc: MockMvc

    @Mock
    private lateinit var pdfProcessingService: PdfProcessingService

    @Mock
    private lateinit var offerDataRepository: OfferDataRepository

    @InjectMocks
    private lateinit var uploadController: UploadController

    @BeforeEach
    fun setup() {
        MockitoAnnotations.openMocks(this)
        mockMvc = MockMvcBuilders.standaloneSetup(uploadController).build()
    }

    @Test
    fun `sollte PDF-Datei erfolgreich hochladen`() {
        val file = MockMultipartFile(
            "file",
            "test.pdf",
            MediaType.APPLICATION_PDF_VALUE,
            "PDF content".toByteArray()
        )

        mockMvc.perform(multipart("/api/upload")
            .file(file)
            .param("pagesPerChunk", "5"))
            .andExpect(status().isOk)
            .andExpect(content().string("Datei erfolgreich verarbeitet und Angebote gespeichert."))

        verify(pdfProcessingService).processPdf(file, 5)
    }

    @Test
    fun `sollte Health-Check erfolgreich durchführen`() {
        mockMvc.perform(get("/api/health"))
            .andExpect(status().isOk)
            .andExpect(jsonPath("$.status").value("UP"))
            .andExpect(jsonPath("$.service").value("handzettelki-backend"))
    }

    @Test
    fun `sollte alle Angebote zurückgeben`() {
        val testOffers = mutableListOf<OfferData?>(
            OfferData(id = 1L, productName = "Test Angebot", price = "9.99€"),
            OfferData(id = 2L, productName = "Test Angebot 2", price = "19.99€")
        )

        `when`(offerDataRepository.findAll()).thenReturn(testOffers)

        mockMvc.perform(get("/api/offers"))
            .andExpect(status().isOk)
            .andExpect(jsonPath("$[0].productName").value("Test Angebot"))
            .andExpect(jsonPath("$[1].productName").value("Test Angebot 2"))
    }

    @Test
    fun `sollte Fehler bei PDF-Verarbeitung behandeln`() {
        val file = MockMultipartFile(
            "file",
            "test.pdf",
            MediaType.APPLICATION_PDF_VALUE,
            "PDF content".toByteArray()
        )

        doThrow(RuntimeException("Verarbeitungsfehler"))
            .`when`(pdfProcessingService).processPdf(file, 5)

        mockMvc.perform(multipart("/api/upload")
            .file(file)
            .param("pagesPerChunk", "5"))
            .andExpect(status().isInternalServerError)
            .andExpect(content().string("Fehler bei der Verarbeitung der PDF: Verarbeitungsfehler"))
    }
}